File: 백준/Gold/2493. 탑/탑.java
import java.io.*;
import java.util.*;

public class Main{
    public static void main(String[] args) throws IOException{
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));
        
        int n = Integer.parseInt(br.readLine());
        StringTokenizer st = new StringTokenizer(br.readLine());
        
        Stack<int[]> stack = new Stack<>();
        for(int i=1; i<=n; i++){
            int num = Integer.parseInt(st.nextToken());
            while(!stack.isEmpty()){
                if(stack.peek()[1] >= num){
                    bw.write(stack.peek()[0]+" ");
                    break;
                }
                stack.pop();
            }
            if(stack.isEmpty()){
                bw.write("0 ");
            }
            stack.push(new int[]{i, num});
        }
        bw.flush();
        br.close();
        bw.close();
    }
}

File: 백준/Gold/2493. 탑/탑.py
import sys
input = sys.stdin.readline

n = int(input())
toplist = list(map(int, input().split()))
answer = [0]*n
stack = []

for i in range(len(toplist)):
    while stack: #stack이 비어있지 않은 동안 실행
        if stack[-1][1] > toplist[i]:
            answer[i] = stack[-1][0] + 1 #인덱스 1부터 시작
            break
        else:
            stack.pop()
    stack.append((i, toplist[i])) #현재 탑의 인덱스와 높이 추가
print(*answer)